File: recipes/llm-voice-assistant/android/llm-voice-assistant/src/main/java/ai/picovoice/llmvoiceassistant/RTFProfiler.java
package ai.picovoice.llmvoiceassistant;

public class RTFProfiler {
    private final int sampleRate;
    private double computeSec;
    private double audioSec;
    private double tickSec;

    public RTFProfiler(int sampleRate) {
        this.sampleRate = sampleRate;
        this.computeSec = 0.0;
        this.audioSec = 0.0;
        this.tickSec = 0.0;
    }

    public void tick() {
        this.tickSec = System.nanoTime() / 1e9;
    }

    public void tock(short[] pcm) {
        this.computeSec += (System.nanoTime() / 1e9) - this.tickSec;
        if (pcm != null && pcm.length > 0) {
            this.audioSec += pcm.length / (double) this.sampleRate;
        }
    }

    public double rtf() {
        double rtf = this.computeSec / this.audioSec;
        this.computeSec = 0.0;
        this.audioSec = 0.0;
        return rtf;
    }
}

File: recipes/llm-voice-assistant/ios/LLMVoiceAssistantDemo/RTFProfiler.swift
import Foundation

class RTFProfiler {
    private let sampleRate: Int32
    private var computeSec: Double
    private var audioSec: Double
    private var tickSec: Double

    init(sampleRate: Int32) {
        self.sampleRate = sampleRate
        self.computeSec = 0.0
        self.audioSec = 0.0
        self.tickSec = 0.0
    }

    func tick() {
        self.tickSec = Double(DispatchTime.now().uptimeNanoseconds) / 1e9
    }

    func tock(pcm: [Int16]?) {
        let nowSec = Double(DispatchTime.now().uptimeNanoseconds) / 1e9
        self.computeSec += nowSec - self.tickSec

        if let pcm = pcm, !pcm.isEmpty {
            self.audioSec += Double(pcm.count) / Double(self.sampleRate)
        }
    }

    func rtf() -> Double {
        let rtf = audioSec > 0.0 ? computeSec / audioSec : 0.0
        self.computeSec = 0.0
        self.audioSec = 0.0
        return rtf
    }
}
